mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/MqttFrameStage.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.mqtt.streaming
+package impl
+
+import akka.annotation.InternalApi
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+
+@InternalApi private[streaming] object MqttFrameStage {
+  @tailrec
+  def frames(
+      maxPacketSize: Int,
+      bytesReceived: ByteString,
+      bytesToEmit: List[ByteString]
+  ): Either[IllegalStateException, (immutable.Iterable[ByteString], ByteString)] = {
+    import MqttCodec._
+
+    val i = bytesReceived.iterator
+    val _ = i.drop(1) // Length starts at offset 1
+    i.decodeRemainingLength() match {
+      case Right(remainingLength) =>
+        val headerSize = bytesReceived.size - i.len
+        val packetSize = remainingLength + headerSize
+        if (packetSize <= maxPacketSize) {
+          if (bytesReceived.size >= packetSize) {
+            val (b0, b1) = bytesReceived.splitAt(packetSize)
+            frames(maxPacketSize, b1, bytesToEmit :+ b0)
```

Comment:
appending to a list is not a fast operation. May I suggest switching to a Vector?

Reply:
You may! Done. https://www.youtube.com/watch?v=fVq4_HhBK8Y

Reply:
:D